dev/src/index.ts

```diff
@@ -502,30 +481,55 @@ export class Firestore implements firestore.Firestore {
       this._projectId = settings.projectId;
     }
 
-    if (settings.host !== undefined) {
+    let url: URL | null = null;
+
+    // If the environment variable is set, it should always take precedence
+    // over any user passed in settings.
+    if (process.env.FIRESTORE_EMULATOR_HOST) {
+      validateHost(
+        'FIRESTORE_EMULATOR_HOST',
+        process.env.FIRESTORE_EMULATOR_HOST
+      );
+
+      settings = {
+        ...settings,
+        host: process.env.FIRESTORE_EMULATOR_HOST,
+        ssl: false,
+      };
+      url = new URL(`http://${settings.host}`);
+    } else if (settings.host !== undefined) {
       validateHost('settings.host', settings.host);
-      if (settings.servicePath !== undefined) {
-        throw new Error(
-          'Cannot set both "settings.host" and "settings.servicePath".'
-        );
-      }
-      if (settings.apiEndpoint !== undefined) {
-        throw new Error(
-          'Cannot set both "settings.host" and "settings.apiEndpoint".'
+      url = new URL(`http://${settings.host}`);
+      if (
+        (settings.servicePath !== undefined &&
+          settings.servicePath !== url.hostname) ||
+        (settings.apiEndpoint !== undefined &&
+          settings.apiEndpoint !== url.hostname)
+      ) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          `The provided host (${url.hostname}) in "settings" does not ` +
+            `match the existing host (${
+              settings.servicePath ?? settings.apiEndpoint
+            }). Using the provided host.`
         );
       }
+    }
 
-      const url = new URL(`http://${settings.host}`);
+    // Only store the host if a valid value was provided in `host`.
+    if (url !== null) {
       settings.servicePath = url.hostname;
       if (url.port !== '' && settings.port === undefined) {
         settings.port = Number(url.port);
       }
-      // We need to remove the `host` setting, in case a user calls `settings()`,
-      // which will again enforce that `host` and `servicePath` are not both
-      // specified.
-      delete settings.host;
     }
 
+    // We need to remove the `host` and `apiEndpoint` setting, in case a user
```

Comment:
Comment no longer applicable.

Reply:
Isn't this still applicable if the user does not provide the env var? In that case, creating a new Firestore with a host passed in and then calling `settings()` with `host` specified would cause this situation.